### PR TITLE
[PROCS-3946] Add local process collector for language detection

### PR DESCRIPTION
**Pull request by daniel-taf**
### What does this PR do?

This PR adds a local process collector for WorkloadMeta to be used for language detection when the process checks are running in the core agent. The main benefit is reduced complexity as the gRPC server is no longer needed. It's worth noting that this collector will only work with the Process check enabled for now. However, this is meant to be an intermediary step and should only affect internal clusters. Ideally, we don't want language detection to be always be reliant on live processes. On Linux, once the check migration is complete, Language detection should be able to run in the core agent with or without the process check. Windows will still be reliant on the remote process collector as it requires elevated permissions for process collection that the Core Agent does not have.

### Motivation

PROCS-3946

### Additional Notes

e2e language detection tests will be updated as a follow up. Another additional piece of work is allowing this to work without live processes enabled on the core agent.

### Describe how to test/QA your changes
Run process checks on the core agent and enabled language detection
```yaml
# datadog.yaml
process_config:
  process_collection:
    enabled: true
  run_in_core_agent:
    enabled: true
language_detection:
  enabled: true
```
Start a python process and verify language is detected with `local_process_collector` as the source.
```bash
$ echo -e 'import time\ntime.sleep(60)' > prog.py
$ nohup python3 prog.py >myscript.log 2>&1 </dev/null &
[1] 13093
$ sudo /opt/datadog-agent/bin/agent/agent workload-list
=== Entity process sources(merged):[local_process_collector] id: 13093 ===
----------- Entity ID -----------
PID: 13093
Namespace PID: 13093
Container ID:
Creation time: 2024-06-18 13:34:48 -0700 PDT
Language: python
===
```

### CODE REVIEW

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:103]: stream method should not be returning anything. 
it should only use channels and logging.

Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go:103]: you can check the [stream method of the docker collector](https://github.com/DataDog/datadog-agent/blob/433dfce57428d08591e1ab7d2a497c8752ad9fa7/comp/core/workloadmeta/collectors/internal/docker/docker.go#L140-L178)

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:114]: no need to return `nil`. 

A simple `return` is enough to stop the go routine.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:117]: I think `CollectorEventsFromProcessDiff` doesn't need to be exported.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:67]: Wouldn't it be simpler to pass the diff itself as an argument here? 

I think it would be more aligned with the function name. 

I don't have a strong opinion though, both should work fine.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Done

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Done

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Agreed. Done

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Done

File: comp/core/workloadmeta/collectors/internal/process/process_collector_test.go
Comment (daniel-taf) [inline comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:49]: I saw this was added in other tests but not 100 percent sure why. Maybe store initialization takes time?

**Review by hestonhoffman — APPROVED** [review]

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:100]: The enablement logic is a bit complex, but it should be temporary until we can collect processes on the core agent without live processes.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:7]: [minor] You can just make this a single line

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:39]: [minor]
- For these private functions, let's put it towards the bottom of the file. 
- Lets add documentation on what this does
- Maybe a better name for this method would be `transform`. The args and return type will implicitly tell the user what this transforms from `ProcessCacheDiff` to `CollectorEvent`

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:75]: docker collector? bad copy/paste?

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:95]: [minor] Is this first/outer paren actually needed?
```
	processChecksInCoreAgent := (
```

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Done

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: done.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: fixed

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: I thought it was for it to properly capture the next line but I suppose the operator is enough!

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:83]: We don't need to do anything here?

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:30]: Can you add a compile time check to verify collector adhere's to the Collector interface?

File: comp/core/workloadmeta/collectors/internal/process/process_collector_test.go
Comment (wiyu) [reply comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:49]: does it work if you don't have a sleep?

File: comp/core/workloadmeta/collectors/internal/process/process_collector_test.go
Comment (wiyu) [inline comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:133]: how about moving this defer to where you create the cancel context instead?

File: pkg/process/metadata/workloadmeta/extractor.go
Comment (wiyu) [inline pkg/process/metadata/workloadmeta/extractor.go:75]: What makes this shared?

**Review by wiyu — APPROVED** [review]
Looks good to me overall, just left some minor comments/questions

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go:83]: I believe just consuming from the channel is all thats needed. Other collectors do the same.

File: pkg/process/metadata/workloadmeta/extractor.go
Comment (daniel-taf) [reply pkg/process/metadata/workloadmeta/extractor.go:75]: Yes, should the naming be different?

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go:30]: Doesn't this [line](https://github.com/DataDog/datadog-agent/blob/2c7e2575cecfb9465bcc8102f805e6305109615d/comp/core/workloadmeta/collectors/internal/process/process_collector.go#L41) already cover that or do I need separate a check like this:
```go
var _ workloadmeta.Collector = &collector{}
```

File: pkg/process/metadata/workloadmeta/extractor.go
Comment (wiyu) [reply pkg/process/metadata/workloadmeta/extractor.go:75]: No, just wasn't obviously to me what this was doing. This can't just be managed by fx right?

**Comment by adel121** [comment]
❓ Question

> It's worth noting that this collector will only work with the live processes enabled for now. 

If I understand correctly, after this PR is merged, running language detection will require live processes collection to be enabled.

If my understanding is correct, any reason why it is done this way? 

My main concern is that live process collection is not available for all customers, and should not be enabled by default. Therefore, I would ideally expect to have a way to run language detection in the core agent without needing to enable live process collection.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:58]: why does live process collection need to be enabled? 

The process-agent version of the process collector doesn't rely on live process collection to enable the collector. It only relies on language detection. (Check [here](https://github.com/DataDog/datadog-agent/blob/201b789380d9c9faac6339d714aced6f0f4a45d8/pkg/process/metadata/workloadmeta/collector/process.go#L120-L133))

It might even cause some issues if a user has the following:
- language detection enabled
- live process collection enabled
- process config run in core agent set to true

I don't think there will be anything that stops having both the process collector and process check running simultaneously on the core agent.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:68]: IMHO, the error message here might be misleading. If language detection was enabled but live process collection / running in core agent is disabled, it will log that language detection is disabled, which is not the case. 

WDYT?

File: comp/core/workloadmeta/collectors/internal/process/process_collector_test.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:133]: Done

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Updated the error message to "language detection or core agent process collection is disabled". Let me know what you think.

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go:60]: We discussed this off Github. This PR is meant to serve as an intermediary step. Its the easiest case to support language detection when the process check is enabled which is why it was tackled first. As a follow-up, we can work on adding the ability to use this without requiring the process check. Worth noting that even then, the remote process collector will still be required on Windows for the foreseeable future as process collection on Windows requires elevated privileges that the core agent does not have.

As for the case where both collectors start, leaving it as it is won't cause issues other than constant retries to begin the stream. To handle it, we can have a check like this:
```go
processChecksInCoreAgent := config.Datadog().GetBool("process_config.process_collection.enabled") &&
		config.Datadog().GetBool("process_config.run_in_core_agent.enabled")
langDetectionEnabled := config.Datadog().GetBool("language_detection.enabled")

return langDetectionEnabled && !processChecksInCoreAgent
```
It may worth having an OS check for Windows to help handle misconfigurations as well. What do you think?

File: comp/core/workloadmeta/collectors/internal/process/process_collector_test.go
Comment (daniel-taf) [reply comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:49]: It does work, but maybe it was there to prevent flakes? @adel121 do you have an idea here?

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go
Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go:60]: Thanks for clarifying.

> It may worth having an OS check for Windows to help handle misconfigurations as well (i.e. someone enables the run in core agent config on Windows). What do you think?

I think it depends on how you wish to expose the feature. 
Are we going to publicly document this and expose it via helm values.yaml for example? 
If yes, then I don't think we need to handle this misconfiguration option because we can simply indicate in the documentation that this option is not supported for windows. If a user tries it on windows and it doesn't work, then it is expected, based on the documentation. 

If no (i.e. if we need to enable/disable this option behind the scenes), then I think we should have an os check somewhere to make surely it is properly configured always.

WDYT?

**Pull request closed without merging**
